### PR TITLE
PTX-2860 Implement method for STC upgrade/validate and add first test

### PR DESCRIPTION
**Pull request by nikolaypopov**
Implementing upgrade/validate methods and adding upgrade test.

1) Implement method for STC upgrade/validate
2) Add first test
3) Make a fix to sched-ops and merged it into operator

Signed-off-by: nikolaypopov <nikolay.popov86@gmail.com>

### CODE REVIEW

File: pkg/util/test/util.go [hunk 1]
```diff
@@ -330,6 +365,57 @@ func ValidateStorageCluster(
...
+
+		// Check StorageNodes status and PX version
+		expectedStatus := "Online"
+		var readyNodes int
+		for _, storageNode := range storageNodeList.Items {
+			if storageNode.Status.Phase == expectedStatus && strings.Contains(storageNode.Spec.Version, pxVersion) {
+				readyNodes++
+				logrus.Debugf("storagenode: %s, Expected status: %s Got: %s, Expected PX version: %s Got: %s", storageNode.Name, expectedStatus, storageNode.Status.Phase, pxVersion, storageNode.Spec.Version)
+			} else {
+				logrus.Debugf("storagenode: %s Expected status: %s Got: %s, Expected PX version: %s Got: %s", storageNode.Name, expectedStatus, storageNode.Status.Phase, pxVersion, storageNode.Spec.Version)
```
Comment (piyush-nimbalkar) [inline pkg/util/test/util.go:402]: Both the debug statements are identical. We can just remove the else and put it after the if condition.

File: pkg/util/test/util.go [hunk 2]
```diff
@@ -50,6 +51,35 @@ import (
...
+	// DefaultValidateDeployRetryInterval is a default retry interval for deployment validation
+	DefaultValidateDeployRetryInterval = 30 * time.Second
+	// DefaultValidateUpgradeTimeout is a default timeout for upgrade validation
+	DefaultValidateUpgradeTimeout = 1400 * time.Second
+	// DefaultValidateUpgradeRetryInterval is a default retry interval for upgrade validation
+	DefaultValidateUpgradeRetryInterval = 60 * time.Second
+	// DefaultValidateUninstallTimeout is a default timeout for uninstall validation
+	DefaultValidateUninstallTimeout = 900 * time.Second
+	// DefaultValidateUninstallRetryInterval is a default retry interval for uninstall validation
+	DefaultValidateUninstallRetryInterval = 30 * time.Second
```
Comment (piyush-nimbalkar) [inline pkg/util/test/util.go:80]: Why not leave these constants in the integration_test package? Isn't it better to have them in the caller package?

File: test/integration_test/common_test.go [hunk 3]
```diff
@@ -135,21 +116,69 @@ func constructStorageCluster() (*corev1.StorageCluster, error) {
...
 	}
 	return cluster, nil
 }
 
-func createStorageCluster(cluster *corev1.StorageCluster) error {
-	_, err := operator.Instance().CreateStorageCluster(cluster)
-	return err
+func createStorageCluster(cluster *corev1.StorageCluster) (*corev1.StorageCluster, error) {
+	cluster, err := operator.Instance().CreateStorageCluster(cluster)
+	return cluster, err
```
Comment (piyush-nimbalkar) [inline test/integration_test/common_test.go:134]: nit: we can simple return directly
```
return operator.Instance().CreateStorageCluster(cluster)
```

File: test/integration_test/common_test.go [hunk 4]
```diff
@@ -135,21 +116,69 @@ func constructStorageCluster() (*corev1.StorageCluster, error) {
...
+					Value: pxDockerUsername,
+				},
+				{
+					Name:  testutil.PxRegistryPasswordEnvVarName,
+					Value: pxDockerPassword,
+				},
+			}
+			cluster.Spec.CommonConfig.Env = append(cluster.Spec.CommonConfig.Env, newEnvVar...)
+		}
+	}
```
Comment (piyush-nimbalkar) [inline test/integration_test/common_test.go:174]: Can we extract these lines into a method that can be used by constructStorageCluster as well?

File: test/integration_test/common_test.go [hunk 5]
```diff
@@ -135,21 +116,69 @@ func constructStorageCluster() (*corev1.StorageCluster, error) {
...
+
+	cluster, err = operator.Instance().UpdateStorageCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return cluster, nil
```
Comment (piyush-nimbalkar) [inline test/integration_test/common_test.go:181]: nit: could just directly return this 
```
return operator.Instance().UpdateStorageCluster(cluster)
```

File: pkg/util/test/util.go (hunk 1 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Done

File: pkg/util/test/util.go (hunk 2 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Moved back to common.go

File: test/integration_test/common_test.go (hunk 4 above)
Comment (nikolaypopov) [reply test/integration_test/common_test.go]: Created a func out of code block.

File: test/integration_test/common_test.go (hunk 3 above)
Comment (nikolaypopov) [reply test/integration_test/common_test.go]: Done

File: test/integration_test/common_test.go (hunk 5 above)
Comment (nikolaypopov) [reply test/integration_test/common_test.go]: Done

File: test/integration_test/common_test.go [hunk 6]
```diff
@@ -85,71 +87,82 @@ func setup() error {
...
+		if err != nil {
+			return err
+		}
+
+		envVarList := []v1.EnvVar{}
+
+		// Add release manifest URL to Env Vars
+		envVarList = append(envVarList, []v1.EnvVar{{Name: testutil.PxReleaseManifestURLEnvVarName, Value: releaseManifestURL}}...)
+
+		// Add Portwrox Docker credentials to Env Vars
```
Comment (piyush-nimbalkar) [inline test/integration_test/common_test.go:152]: nit: *Portworx

File: pkg/util/test/util.go [hunk 7]
```diff
@@ -50,6 +51,22 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 )
 
+const (
+	// SpecDir is a directory with all the specs
+	SpecDir = "./operator-test"
+
+	// PxNamespace is a default namespace for StorageCluster
+	PxNamespace = "kube-system"
```
Comment (piyush-nimbalkar) [inline pkg/util/test/util.go:59]: Can we move these back as well?

**Review by piyush-nimbalkar — APPROVED** [review]
LGTM
Merge after addressing the nit picks.

File: pkg/util/test/util.go (hunk 7 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Moved.

**Pull request merged**